apps/serdes/Deserializer.cpp

```diff
@@ -0,0 +1,608 @@
+#include "Deserializer.h"
+#include <fstream>
+#include <iostream>
+
+std::string Deserializer::deserialize_string(const flatbuffers::String *str) {
+    return str->str();
+}
+
+Halide::Type Deserializer::deserialize_type(const Halide::Serdes::Type *type) {
+    // bits
+    int bits = type->bits();
+
+    // lanes
+    int lanes = type->lanes();
+
+    // code
+    Halide::Serdes::TypeCode code_deserialized = type->code();
+
+    halide_type_code_t code = halide_type_uint;
+    switch (code_deserialized) {
+    case Halide::Serdes::TypeCode::TypeCode_Int:
+        code = halide_type_int;
+        break;
+    case Halide::Serdes::TypeCode::TypeCode_UInt:
+        code = halide_type_uint;
+        break;
+    case Halide::Serdes::TypeCode::TypeCode_Float:
+        code = halide_type_float;
+        break;
+    case Halide::Serdes::TypeCode::TypeCode_Handle:
+        code = halide_type_handle;
+        break;
+    case Halide::Serdes::TypeCode::TypeCode_BFloat:
+        code = halide_type_bfloat;
+        break;
+    }
+
+    return Halide::Type(code, bits, lanes);
+}
+
+Halide::Internal::Function Deserializer::deserialize_function(const Halide::Serdes::Func *function) {
+    // name
+    std::string name = deserialize_string(function->name());
+
+    // origin_name
+    std::string origin_name = deserialize_string(function->origin_name());
+
+    // output_types
+    std::vector<Halide::Type> output_types;
+    auto output_types_serialized = function->output_types();
+    for (const auto &type : *output_types_serialized) {
+        output_types.push_back(deserialize_type(type));
+    }
+
+    // required_types
+    std::vector<Halide::Type> required_types;
+    auto required_types_serialized = function->required_types();
+    for (const auto &type : *required_types_serialized) {
+        required_types.push_back(deserialize_type(type));
+    }
+
+    // required_dimensions
+    int required_dim = function->required_dims();
+
+    // args
+    std::vector<std::string> args;
+    auto args_serialized = function->args();
+    for (const auto &arg : *args_serialized) {
+        args.push_back(deserialize_string(arg));
+    }
+
+    // assemble the function
+    return Halide::Internal::Function(name, origin_name, output_types, required_types, required_dim, args);
+}
+
+Halide::Internal::Stmt Deserializer::deserialize_stmt(uint8_t type_code, const void *stmt) {
```

Comment:
Seems like the `type_code` argument should really be `Halide::Serdes::TypeCode` instead. Why are we using `uint8` here?